### ISSUE TITLE
Extend instrumentation capabilities.

### ISSUE THREAD

**Issue by lesovsky** [jackc/pgx#782]
Hi,

I've tried to add some Prometheus-based instrumentation into an application that uses `pgxpool` and found lack of things which should help me. In general, I'd like to implement basic metrics such as queries per second and queries durations and then extend metrics with driver-specific things based on `Stat` struct. But only one thing I could made is a just transactions per second (not queries).

It would be great if functionality of pgxpool will be extended in the following way:
1. Context can be transferred from `BeforeAcquire` to `AfterRelease` function - it will make possible to measure time of connection were used.
2. Return to `AfterRelease` some information about executed query, such as error/success status, bytes sent/received, text of the statement may be, something else...
3. Inside of `AfterRelease`, allow to get access to pool's `Stat` struct - it may helps to expose driver-specific stats into metrics.

I consider these things may not be easy to implement, but it can be very useful for application-side instrumentation.

Also, I'd like to mention an approach used in [Redis driver](https://github.com/go-redis/redis). It uses [Hook](https://github.com/go-redis/redis/blob/master/redis.go#L22) interface which provides `BeforeProcess` and `AfterProcess` methods which can be used to get information about executed commands (from `Cmder` struct). Using this methods I made verbose metrics that helps to understand Redis workload from application's point of view.

Regards, Alexey

**Comment by jackc** [comment]
These are all potentially valuable changes, but difficult or impossible to do without breaking backwards compatibility.

> Context can be transferred from BeforeAcquire to AfterRelease function - it will make possible to measure time of connection were used.

pgx and pgxpool generally only use a context for a single action (e.g. A context passed to `Begin()` is only used for the `begin;` query not for the entire transaction). This may or may not have been the ideal design, but it is baked in pretty deep at this point and changing it would be a breaking change. Changing the signature of `AfterRelease()` would obviously also be a breaking change.

> Return to AfterRelease some information about executed query, such as error/success status, bytes sent/received, text of the statement may be, something else...

A connection may have executed multiple queries while checked out. Query metrics might be better handled by a wrapper around `*pgx.Conn` or `pgxpool.Conn`.

It _may_ make sense to fork `pgxpool` for your own use. Part of the design goals for `v4` of `pgx` was to allow alternative pool implementations. The main `pgx` library has no dependencies on `pgxpool` so a different pool library can be used without any changes to the pgx library. In addition, the tricky concurrency parts of a connection pool are handled in https://github.com/jackc/puddle. The pgxpool package itself is almost entirely simple wrappers and glue code -- so you could probably reshape it as desired without substantial risk or effort.

**Comment by lesovsky** [comment]
> but difficult or impossible to do without breaking backwards compatibility.

Yes, I thought in the same way and not expect quick resolution. Basically, this issue can be considered as a feature request that can be implemented in the future.

> Query metrics might be better handled by a wrapper around *pgx.Conn or pgxpool.Conn.

This is the way I currently trying to implement in my application.

**Comment by atombender** [comment]
I'd like better support for this. In particular, right now it's impossible to build a Prometheus histogram for connection acquisition time. All you have is `stats.AcquireDuration()`, which is the cumulative time; so we have no choice but to monitor the stats for this.

A hook interface would be nice indeed. Backwards compatibility could be maintained by adding a hook interface while keeping the old existing hooks.
